File: src/main/java/com/vc/medium/L90.scala
package com.vc.medium

object L90 {
  def subsetsWithDup(arr: Array[Int]): List[List[Int]] = {
    import scala.collection.mutable
    val res = new mutable.ListBuffer[List[Int]]()
    val list = new mutable.ArrayBuffer[Int]()
    val n = arr.length
    scala.util.Sorting.quickSort(arr)
    def solve(start: Int): Unit = {
      res += list.toList
      (start until n).foreach(i => {
        if(!(i > start && arr(i) == arr(i - 1))) {
          list += arr(i)
          solve(i + 1)
          list.remove(list.size - 1)
        }
      })
    }
    solve(0)
    res.toList
  }
}

File: src/main/java/com/vc/medium/L90.java
package com.vc.medium;

import java.util.*;

class L90 {
    List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> subsetsWithDup(int[] arr) {
        boolean[] visited = new boolean[arr.length];
        Arrays.sort(arr);
        solve(0, new ArrayList<Integer>(), arr, visited);
        return res;
    }

    private void solve(int index, List<Integer> list, int[] arr, boolean[] visited) {
        res.add(new ArrayList<>(list));
        for(int i = index; i < arr.length; i++) {
            if(visited[i]) continue;
            if(i > 0 && arr[i] == arr[i - 1] && !visited[i - 1]) continue;

            list.add(arr[i]);
            visited[i] = true;
            solve(i + 1, list, arr, visited);
            visited[i] = false;
            list.remove(list.size() - 1);
        }
    }
}
